juju/juju
# [JUJU-1857] Discover manual cloud subnets on loopback device

For manual clouds, we discover subnets based on machine NIC addresses as they are added. We have to-date been skipping loopback devices for this behaviour.

A requirement now arises to discover /32 subnets added to loopback devices on a cloud using BGP routing.

Here we change the discovery logic to include the loopback devices, instead omitting the subnets for addresses scoped to the local machine, such as 127.0.0.1/8.

To achieve this, a change to ensure that address scope is passed with discovered network configuration is included. This does not change data saved to Mongo, and does not require an upgrade step.

## QA steps

On any machine that can be used as manually-provisioned:
- SSH to the machine before provisioning and run `sudo ip addr add 10.1.1.2/32 dev lo`.
- Then either bootstrap a manual cloud onto this machine, or add it to an extant manual cloud.
- After it comes up, we should see the CIDR listed in `juju subnets` output:
```
subnets:
  10.1.1.2/32:
    type: ipv4
    status: in-use
    space: alpha
    zones: []
  172.16.99.0/24:
    type: ipv4
    status: in-use
    space: alpha
    zones: []
```

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1990661
